### ISSUE TITLE
How can we use ScalaCheck to return Arbitrary[T] from pre-canned Seq[T]?

### ISSUE THREAD

**Issue by fommil** [rickynils/scalacheck#137]
**TL;DR**: `Gen.oneOf` is really slow for big collections.

I am upgrading from scalacheck 1.9 to 1.12.1 and some of my code is not compiling anymore.

I have two ScalaCheck Arbitrary[MyData] implementations: one that uses fake generators and another that uses a pre-canned sample that was created from our database. I therefore have two tests that run: one on "fake" data and one on "real" data.

The "real" Arbitrary implementation used to look like this (sorry for total misuse of your API):

``` scala
implicit val arbMyData: Arbitrary[MyData] = Arbitrary {
    var stream: Stream[MyData] = importer.parse()
    Gen { params => // ignores params
      val head = stream.headOption
      if (stream.nonEmpty)
        stream = stream.tail
      head
    }
  }
```

where `stream` can potentially contain billions of entries (if the developer wants to use a custom sample if they are specifically working in this area). BTW, we have a way of pre-calculating the size of the dataset and call it with 

``` scala
    forAll(minSuccessful(size)) { (d: MyData) =>
      ...
    }
```

(otherwise we only get 100 samples)

Rewriting the arb in 1.12 as

``` scala
  implicit val arbMyData: Arbitrary[MyData] = Arbitrary {
    oneOf(importer.parse())
  }
```

is **incredibly** slow, even for small datasets of 10,000 entries.  We tried using importer.parse().toVector() but it didn't change anything, and even assigning the vector to a val outside the gen resulted in a lot of CPU time inside `SeqLike.contains`.

We came up with this hack

``` scala
  val entries = importer.parse().toVector
  implicit val arbMyData: Arbitrary[MyData] = Arbitrary {
    Gen.choose(0, entries.size - 1).map(index => entries(index))
  }
```

but it isn't ideal because it doesn't guarantee that we'll get every entry exactly once if we ask for `entries.size` values.

What is the best way to return an Arbitrary[T] from an existing Seq[T] and guarantee that everything is returned exactly once?

**Comment by rickynils** [comment]
Interesting use case! I can understand that the (admittedly hacky) generator filtering introduced in 1.11 slows things down for you.

However, you can still use a similar solution as before. For example, you can use `Gen.wrap` which creates a generator that evaluates another generator. Something like this would work:

```
val genData: Gen[Int] = {
  var stream = Stream(1,2,3)
  Gen.wrap {
    stream.headOption match {
      case None => Gen.fail
      case Some(n) =>
        stream = stream.tail
        Gen.const(n)
    }
  }
}
```

As you can see, this generator will be exhausted after 3 evaluations:

```
scala> genData.sample
res10: Option[Int] = Some(1)

scala> genData.sample
res11: Option[Int] = Some(2)

scala> genData.sample
res12: Option[Int] = Some(3)

scala> genData.sample
res13: Option[Int] = None

scala> genData.sample
res14: Option[Int] = None
```

In ScalaCheck 2, this kind of generator exhaustion will be supported without introducing mutable state within the generator. `Gen.oneOf` should then be able to work as you expect (and I will try to take care to make it work nicely with large data sets).

**Comment by skarepov** [comment]
@rickynils Many thanks, it works again.

**Comment by yaravind** [comment]
I was looking for this functionality (generated data based on reference data). However, is it possible to avoid duplicates during the data generation? I.e. the Gens should always return distinct items
